scalanet/src/io/iohk/scalanet/peergroup/InMemoryPeerGroup.scala

```diff
@@ -0,0 +1,142 @@
+package io.iohk.scalanet.peergroup
+
+import monix.eval.Task
+import monix.reactive.{MulticastStrategy, Observable}
+import monix.reactive.subjects.ConcurrentSubject
+import monix.execution.Scheduler.Implicits.global
+import InMemoryPeerGroup._
+
+import scala.collection.concurrent.TrieMap
+import scala.util.Random
+
+// For simplicity we can model the network with three main interactions.
+// 1. Register to the network.
+// 2. Disconnect from the network.
+// 3. Send messages.
+class Network[A, M] {
+  private val peers: TrieMap[A, InMemoryPeerGroup[A, M]] = TrieMap()
+
+  def clear(): Unit = peers.clear()
+
+  def register(peer: InMemoryPeerGroup[A, M]): Result[Unit] =
+    if (peers.get(peer.processAddress).isEmpty) {
+      peers += (peer.processAddress -> peer)
+      allGood(())
+    } else error("The address is already in use")
+
+  def disconnect(peer: InMemoryPeerGroup[A, M]): Unit = peers -= peer.processAddress
+
+  def deliverMessage(channelID: ChannelID, from: A, to: A, message: M): Result[Unit] =
+    peers.get(to) match {
+      case None => error(s"Unreachable peer $to")
+      case Some(destination) =>
+        allGood(destination.receiveMessage(channelID, from, message))
+    }
+}
+
+class InMemoryPeerGroup[A, M](address: A)(implicit network: Network[A, M]) extends PeerGroup[A, M] {
+
+  private var status: PeerStatus = PeerStatus.NotInitialized
+  private val channelStream = ConcurrentSubject[InMemoryChannel[A, M]](MulticastStrategy.publish)
+  private val channelsMap: TrieMap[ChannelID, InMemoryChannel[A, M]] = TrieMap()
+  def receiveMessage(channelID: ChannelID, from: A, msg: M): Unit = {
```

Comment:
If this is not supposed to go in the public interface, then make it private.

Reply:
I will check if I can make it only accessible to the Network class.